infra/src/main/kotlin/fi/fta/geoviite/infra/integration/CalculatedChangesService.kt

```diff
@@ -267,52 +271,102 @@ class CalculatedChangesService(
         }
+
+    private fun getAllSwitchJointChanges(
+        locationTrack: LocationTrack,
+        fetchSwitchById: (id: IntId<LayoutSwitch>) -> LayoutSwitch?,
+        getGeocodingContext: (id: IntId<LayoutTrackNumber>) -> GeocodingContext?,
+    ): List<Pair<IntId<LayoutSwitch>, List<SwitchJointDataHolder>>> {
+        val alignment = alignmentDao.fetch(requireNotNull(locationTrack.alignmentVersion))
+        val trackNumberId = locationTrack.trackNumberId
+        val geocodingContext = getGeocodingContext(trackNumberId)
+
+        return geocodingContext?.let { context ->
+            getSwitchJointChanges(
+                locationTrack = locationTrack,
+                alignment = alignment,
+                geocodingContext = context,
+                fetchSwitch = fetchSwitchById,
+                fetchStructure = switchLibraryService::getSwitchStructure,
+            )
+        } ?: emptyList()
+    }
+
+    fun getSwitchChangesFromChangedLocationTrackKms(
+        fetchLocationTrackById: (id: IntId<LocationTrack>) -> LocationTrack?,
+        fetchSwitchById: (id: IntId<LayoutSwitch>) -> LayoutSwitch?,
+        getGeocodingContext: (id: IntId<LayoutTrackNumber>) -> GeocodingContext?,
+        locationTrackId: IntId<LocationTrack>,
+        filterKmNumbers: Collection<KmNumber>,
+        extIds: AllOids,
+    ): List<SwitchChange> {
+        val locationTrack =
+            fetchLocationTrackById(locationTrackId)
+                ?: throw NoSuchEntityException(LocationTrack::class, locationTrackId)
+        val allChanges = getAllSwitchJointChanges(locationTrack, fetchSwitchById, getGeocodingContext)
+        return processSwitchJointChangesByLocationTrackKmChange(
+            allChanges,
+            locationTrackId,
+            locationTrack.trackNumberId,
+            filterKmNumbers,
+            extIds,
+        )
+    }
+
+    fun getSwitchChangesFromChangedLocationTrackKmsByMoment(
+        layoutBranch: LayoutBranch,
+        locationTrackId: IntId<LocationTrack>,
+        moment: Instant,
+        extIds: AllOids,
+        filterKmNumbers: Collection<KmNumber>,
+    ): List<SwitchChange> =
+        getSwitchChangesFromChangedLocationTrackKms(
+            fetchLocationTrackById = { id -> locationTrackService.getOfficialAtMoment(layoutBranch, id, moment) },
+            fetchSwitchById = { id -> switchService.getOfficialAtMoment(layoutBranch, id, moment) },
+            getGeocodingContext = { id -> geocodingService.getGeocodingContextAtMoment(layoutBranch, id, moment) },
+            extIds = extIds,
+            locationTrackId = locationTrackId,
+            filterKmNumbers = filterKmNumbers,
+        )
+
+    fun getSwitchChangesFromChangedLocationTrackKmsByLocationTrackChange(
+        validationVersions: ValidationVersions,
+        locationTrackChange: LocationTrackChange,
+    ): List<SwitchChange> {
+        val validationContext = publicationValidationService.createValidationContext(validationVersions)
```

Comment:
Katsonpa vielä hieman tämän (ja tämän myötä createValidationContext()in julkistamisen) tarpeellisuutta, tämä oli vähän hetkessä tehty juttu missä tuli katsottua että tuollahan tuo menisi kätevästi...